### PR TITLE
Add `BaseException.__new__`

### LINKED ISSUE #12231: `BaseException` lacks `__new__()`
### Summary

When overriding `__new__()` in a `BaseException` subclass, a call of form `super().__new__(cls, arg0, [arg1, [arg2, ...]]])` causes type checkers to complain about unused positional argument(s), even though the positional args are both accepted and used by `BaseException.__new__()` — they become `<BaseException>.args`.

**Root cause:** type checkers believe `super().__new__(...)` in this context is a call to `object.__new__()`, rather than `BaseException.__new__()`.

**Proposed fix:** typeshed should include a definition for `BaseException.__new__()` analogous to and compatible with to its existing signature for `BaseException.__init__()`:
```python3
class BaseException:
    ...
    def __init__(self, *args: object) -> None: ...
    ...
```

### Concrete example

Analyzing the following code....
```python3
from __future__ import annotations

class SomeException(Exception):

    def __new__(cls, alfa: str, bravo: str) -> SomeException:

        msg = f"{alfa} ({bravo})"
        return super().__new__(cls, msg)  # ᐊ── errors occur here

    def __init__(self, alfa: str, bravo: str):

        # (preserve these as attributes)
        pass
```

...with [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=7c4d5d0361e76f53effba5c2b8f0f476) causes this report:
```
bug.py|8| error: Too many arguments for "__new__" of "object"  [call-arg]
|| Found 1 error in 1 file (checked 1 source file)
```

...while [pyright](https://pyright-play.net/?code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChGBjAG2PvqgGVI0AogA8WaXDToAKYaPG1SASgBczKGqgATNMFgxSaAO5xJ7egBoSbYMSVR6SEBYBGIYgDcwt%2ByAVQAtAB8vPwyYhKkKozq0VAQ9ADmUAC8UMAARADexFbEAL5Qkhku7mC5CmlRMWroKCCkdgi4aCCSCgB0cPpGMCZs5rEJvlAAxFCAUCiAAKQTUM3gINxgLCyoUAAWzWiq6lo6cFikWEjG9GhWFtnWXg7Orh5XPpFVaqOSuOgnIG5oUEjrJyTcagOLBOZBoegKSpVXCcehAA) (without `strict`) says:
```
|| .../bug.py
bug.py|8| 37 - error: Expected 1 positional argument (reportCallIssue)
|| 1 error, 0 warnings, 0 informations 
```

### PR DESCRIPTION
**Pull request by tomasr8**
Closes #12231

I think this is the correct signature given the definition:

https://github.com/python/cpython/blob/65fededf9cc1780d5edbef8a6e0a7cf9bc15aea6/Objects/exceptions.c#L43-L69

### CODE REVIEW

**Comment by srittau** [comment]
It seems that pyright does not like the mismatch between `__init__` and `__new__` in sub-classes of `BaseException`. I'm not sure what to do about that, to be honest. Considering that `BaseException_new` just ignores keyword arguments, we could try adding a `kwds: Unused` argument to `BaseException.__new__`.

**Comment by tomasr8** [comment]
That seems to reduce the amount the errors but a lot of them still remain (I pushed the change so that you can see for yourself)

**Comment by srittau** [comment]
I'm a bit unsure how to proceed here. The stubs now seem to match the implementation. I'm not sure why pyright complains.

**Comment by tomasr8** [comment]
I'm as stumped as you are :sweat_smile:

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -1868,6 +1869,7 @@ class BaseException:
     __suppress_context__: bool
     __traceback__: TracebackType | None
     def __init__(self, *args: object) -> None: ...
+    def __new__(cls, *args: object, **kwds: Unused) -> Self: ...
```
Comment (JelleZijlstra) [inline stdlib/builtins.pyi:1872]:
```suggestion
    def __new__(cls, *args: Any, **kwds: Any) -> Self: ...
```
Pyright is enforcing this section of the spec: https://typing.readthedocs.io/en/latest/spec/constructors.html#consistency-of-new-and-init. The signatures of `__new__` and `__init__` need to be consistent, i.e. they need to be able to materialize to the same static type. Therefore, we need to use an Any/Any signature, which materializes to any signature.

Pyright's error message I think has room for improvement here, so I'll report a bug about that.

**Review by srittau — APPROVED** [review]
Thanks, @tomasr8 and @JelleZijlstra!

**Pull request merged**